File: tools/artifactctl/cmd/artifactctl/fixture_image.go
package main

import (
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/poruru-code/esb/pkg/artifactcore"
	proxybuildargs "github.com/poruru-code/esb/pkg/proxy/buildargs"
)

const (
	fixtureImageEnsureSchemaVersion = 1
	defaultFixtureImageRoot         = "e2e/fixtures/images/lambda"
	javaFixtureName                 = "esb-e2e-image-java"
	javaFixtureMavenBaseImage       = "public.ecr.aws/sam/build-java21@sha256:5f78d6d9124e54e5a7a9941ef179d74d88b7a5b117526ea8574137e5403b51b7"
)

var (
	localFixtureSubdirs = map[string]string{
		"esb-e2e-image-python": "python",
		"esb-e2e-image-java":   "java",
	}
	fixtureDockerfileFromPattern = regexp.MustCompile(`(?i)^FROM(?:\s+--platform=[^\s]+)?\s+([^\s]+)`) //nolint:lll
	javaFixtureMavenArgPattern   = regexp.MustCompile(`(?im)^\s*ARG\s+MAVEN_IMAGE(?:\s*=.*)?\s*$`)
	javaFixtureFromPattern       = regexp.MustCompile(`(?im)^\s*FROM\s+\$\{?MAVEN_IMAGE\}?\s+AS\s+builder\s*$`)
)

type fixtureCommandRunner interface {
	Run(cmd []string) error
}

type FixtureImageEnsureInput struct {
	ArtifactPath    string
	NoCache         bool
	FixtureRoot     string
	Runner          fixtureCommandRunner
	EnsureMavenShim func(MavenShimEnsureInput) (MavenShimEnsureResult, error)
	LogWriter       io.Writer
}

type FixtureImageEnsureResult struct {
	SchemaVersion  int      `json:"schema_version"`
	PreparedImages []string `json:"prepared_images"`
}

func executeFixtureImageEnsure(input FixtureImageEnsureInput) (FixtureImageEnsureResult, error) {
	return executeFixtureImageEnsureWithLogWriter(input, os.Stderr)
}

func executeFixtureImageEnsureWithLogWriter(
	input FixtureImageEnsureInput,
	logWriter io.Writer,
) (FixtureImageEnsureResult, error) {
	manifestPath := strings.TrimSpace(input.ArtifactPath)
	if manifestPath == "" {
		return FixtureImageEnsureResult{}, fmt.Errorf("artifact manifest path is empty")
	}
	manifest, err := artifactcore.ReadArtifactManifest(manifestPath)
	if err != nil {
		return FixtureImageEnsureResult{}, fmt.Errorf("read artifact manifest: %w", err)
	}
	sources, err := collectLocalFixtureImageSources(manifest, manifestPath)
	if err != nil {
		return FixtureImageEnsureResult{}, err
	}
	if len(sources) == 0 {
		return FixtureImageEnsureResult{
			SchemaVersion:  fixtureImageEnsureSchemaVersion,
			PreparedImages: []string{},
		}, nil
	}

	runner := input.Runner
	if runner == nil {
		writer := logWriter
		if input.LogWriter != nil {
			writer = input.LogWriter
		}
		runner = stderrCommandRunner{writer: writer}
	}
	ensureMavenShim := input.EnsureMavenShim
	if ensureMavenShim == nil {
		writer := logWriter
		if input.LogWriter != nil {
			writer = input.LogWriter
		}
		ensureMavenShim = func(in MavenShimEnsureInput) (MavenShimEnsureResult, error) {
			return executeMavenShimEnsureWithLogWriter(in, writer)
		}
	}

	fixtureRoot := strings.TrimSpace(input.FixtureRoot)
	if fixtureRoot == "" {
		fixtureRoot = defaultFixtureImageRoot
	}
	fixtureRootAbs, err := filepath.Abs(fixtureRoot)
	if err != nil {
		return FixtureImageEnsureResult{}, fmt.Errorf("resolve fixture root: %w", err)
	}

	preparedImages := make([]string, 0, len(sources))
	resolvedShimByRegistry := map[string]string{}
	for _, source := range sources {
		fixtureName := fixtureRepoName(source)
		subdir, ok := localFixtureSubdirs[fixtureName]
		if !ok {
			return FixtureImageEnsureResult{}, fmt.Errorf("unknown local fixture image source: %s", source)
		}
		fixtureDir := filepath.Join(fixtureRootAbs, subdir)
		if info, statErr := os.Stat(fixtureDir); statErr != nil || !info.IsDir() {
			if statErr != nil {
				return FixtureImageEnsureResult{}, fmt.Errorf("local fixture image source not found: %s", fixtureDir)
			}
			return FixtureImageEnsureResult{}, fmt.Errorf("local fixture image source is not a directory: %s", fixtureDir)
		}

		buildArgs := map[string]string{}
		if fixtureName == javaFixtureName {
			if err := assertJavaFixtureUsesMavenShimContract(filepath.Join(fixtureDir, "Dockerfile")); err != nil {
				return FixtureImageEnsureResult{}, err
			}
			registryHost := imageRegistryHost(source)
			cacheKey := javaFixtureMavenBaseImage + "|" + registryHost
			shimImage, ok := resolvedShimByRegistry[cacheKey]
			if !ok {
				shimResult, shimErr := ensureMavenShim(MavenShimEnsureInput{
					BaseImage:    javaFixtureMavenBaseImage,
					HostRegistry: registryHost,
					NoCache:      input.NoCache,
				})
				if shimErr != nil {
					return FixtureImageEnsureResult{}, fmt.Errorf("maven shim ensure failed: %w", shimErr)
				}
				shimImage = strings.TrimSpace(shimResult.ShimImage)
				if shimImage == "" {
					return FixtureImageEnsureResult{}, fmt.Errorf("maven shim ensure returned empty shim image")
				}
				resolvedShimByRegistry[cacheKey] = shimImage
			}
			buildArgs["MAVEN_IMAGE"] = shimImage
		}

		buildCmd := buildxBuildCommandForFixture(source, fixtureDir, input.NoCache, buildArgs)
		if runErr := runner.Run(buildCmd); runErr != nil {
			return FixtureImageEnsureResult{}, fmt.Errorf("build local fixture image %s: %w", source, runErr)
		}
		if runErr := runner.Run([]string{"docker", "push", source}); runErr != nil {
			return FixtureImageEnsureResult{}, fmt.Errorf("push local fixture image %s: %w", source, runErr)
		}
		preparedImages = append(preparedImages, source)
	}

	return FixtureImageEnsureResult{
		SchemaVersion:  fixtureImageEnsureSchemaVersion,
		PreparedImages: preparedImages,
	}, nil
}

func buildxBuildCommandForFixture(
	tag string,
	contextDir string,
	noCache bool,
	buildArgs map[string]string,
) []string {
	cmd := []string{
		"docker",
		"buildx",
		"build",
		"--platform",
		"linux/amd64",
		"--load",
	}
	if noCache {
		cmd = append(cmd, "--no-cache")
	}
	cmd = proxybuildargs.AppendDockerBuildArgsFromOS(cmd)
	if len(buildArgs) > 0 {
		keys := make([]string, 0, len(buildArgs))
		for key := range buildArgs {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		for _, key := range keys {
			value := strings.TrimSpace(buildArgs[key])
			if value == "" {
				continue
			}
			cmd = append(cmd, "--build-arg", key+"="+value)
		}
	}
	cmd = append(cmd, "--tag", tag, contextDir)
	return cmd
}

func collectLocalFixtureImageSources(
	manifest artifactcore.ArtifactManifest,
	manifestPath string,
) ([]string, error) {
	sources := make(map[string]struct{})
	for i := range manifest.Artifacts {
		artifactRoot, err := manifest.ResolveArtifactRoot(manifestPath, i)
		if err != nil {
			return nil, err
		}
		walkErr := filepath.WalkDir(artifactRoot, func(path string, d fs.DirEntry, walkErr error) error {
			if walkErr != nil {
				return walkErr
			}
			if d.IsDir() {
				return nil
			}
			if d.Name() != "Dockerfile" {
				return nil
			}
			content, readErr := os.ReadFile(path)
			if readErr != nil {
				return readErr
			}
			for _, line := range strings.Split(string(content), "\n") {
				trimmed := strings.TrimSpace(line)
				match := fixtureDockerfileFromPattern.FindStringSubmatch(trimmed)
				if match == nil {
					continue
				}
				source := strings.TrimSpace(match[1])
				if isLocalFixtureImageSource(source) {
					sources[source] = struct{}{}
				}
			}
			return nil
		})
		if walkErr != nil {
			return nil, fmt.Errorf("scan artifact root %s: %w", artifactRoot, walkErr)
		}
	}
	result := make([]string, 0, len(sources))
	for source := range sources {
		result = append(result, source)
	}
	sort.Strings(result)
	return result, nil
}

func isLocalFixtureImageSource(source string) bool {
	if strings.TrimSpace(source) == "" {
		return false
	}
	_, ok := localFixtureSubdirs[fixtureRepoName(source)]
	return ok
}

func fixtureRepoName(source string) string {
	withoutDigest := strings.SplitN(source, "@", 2)[0]
	lastSegment := withoutDigest[strings.LastIndex(withoutDigest, "/")+1:]
	return strings.SplitN(lastSegment, ":", 2)[0]
}

func imageRegistryHost(imageRef string) string {
	withoutDigest := strings.SplitN(strings.TrimSpace(imageRef), "@", 2)[0]
	if !strings.Contains(withoutDigest, "/") {
		return ""
	}
	candidate := strings.SplitN(withoutDigest, "/", 2)[0]
	if candidate == "" {
		return ""
	}
	if candidate == "localhost" || strings.Contains(candidate, ".") || strings.Contains(candidate, ":") {
		return candidate
	}
	return ""
}

func assertJavaFixtureUsesMavenShimContract(dockerfile string) error {
	content, err := os.ReadFile(dockerfile)
	if err != nil {
		return fmt.Errorf("java fixture Dockerfile not found: %s", dockerfile)
	}
	text := string(content)
	if javaFixtureMavenArgPattern.FindStringIndex(text) == nil {
		return fmt.Errorf(
			"java fixture Dockerfile must define `ARG MAVEN_IMAGE` so E2E can inject maven-shim: %s",
			dockerfile,
		)
	}
	if javaFixtureFromPattern.FindStringIndex(text) == nil {
		return fmt.Errorf(
			"java fixture Dockerfile must use `FROM ${MAVEN_IMAGE} AS builder` for proxy-safe Maven resolution: %s",
			dockerfile,
		)
	}
	return nil
}

File: tools/cli/fixture_image.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

from tools.cli import artifact
from tools.cli.common import append_proxy_build_args, run_command
from tools.cli.maven_shim import EnsureInput as MavenShimEnsureInput
from tools.cli.maven_shim import ensure_image as ensure_maven_shim_image

FIXTURE_IMAGE_ENSURE_SCHEMA_VERSION = 1
DEFAULT_FIXTURE_IMAGE_ROOT = "e2e/fixtures/images/lambda"
JAVA_FIXTURE_NAME = "esb-e2e-image-java"
JAVA_FIXTURE_MAVEN_BASE_IMAGE = (
    "public.ecr.aws/sam/build-java21@sha256:"
    "5f78d6d9124e54e5a7a9941ef179d74d88b7a5b117526ea8574137e5403b51b7"
)

_LOCAL_FIXTURE_SUBDIRS: dict[str, str] = {
    "esb-e2e-image-python": "python",
    "esb-e2e-image-java": "java",
}
_DOCKERFILE_FROM_PATTERN = re.compile(r"(?i)^FROM(?:\s+--platform=[^\s]+)?\s+([^\s]+)")
_JAVA_FIXTURE_MAVEN_ARG_PATTERN = re.compile(r"(?im)^\s*ARG\s+MAVEN_IMAGE(?:\s*=.*)?\s*$")
_JAVA_FIXTURE_FROM_PATTERN = re.compile(r"(?im)^\s*FROM\s+\$\{?MAVEN_IMAGE\}?\s+AS\s+builder\s*$")


@dataclass(frozen=True)
class FixtureImageEnsureInput:
    artifact_path: str
    no_cache: bool = False
    fixture_root: str = DEFAULT_FIXTURE_IMAGE_ROOT


@dataclass(frozen=True)
class FixtureImageEnsureResult:
    schema_version: int
    prepared_images: list[str]


def execute_fixture_image_ensure(input_data: FixtureImageEnsureInput) -> FixtureImageEnsureResult:
    manifest_path = input_data.artifact_path.strip()
    if manifest_path == "":
        raise RuntimeError("artifact manifest path is empty")
    manifest = artifact.read_artifact_manifest(manifest_path, validate=True)
    sources = collect_local_fixture_image_sources(manifest, manifest_path)
    if not sources:
        return FixtureImageEnsureResult(
            schema_version=FIXTURE_IMAGE_ENSURE_SCHEMA_VERSION,
            prepared_images=[],
        )

    fixture_root = input_data.fixture_root.strip() or DEFAULT_FIXTURE_IMAGE_ROOT
    fixture_root_abs = Path(fixture_root).resolve()
    prepared_images: list[str] = []
    resolved_shim_by_registry: dict[str, str] = {}

    for source in sources:
        fixture_name = fixture_repo_name(source)
        if fixture_name not in _LOCAL_FIXTURE_SUBDIRS:
            raise RuntimeError(f"unknown local fixture image source: {source}")
        fixture_dir = fixture_root_abs / _LOCAL_FIXTURE_SUBDIRS[fixture_name]
        if not fixture_dir.exists():
            raise RuntimeError(f"local fixture image source not found: {fixture_dir}")
        if not fixture_dir.is_dir():
            raise RuntimeError(f"local fixture image source is not a directory: {fixture_dir}")

        build_args: dict[str, str] = {}
        if fixture_name == JAVA_FIXTURE_NAME:
            assert_java_fixture_uses_maven_shim_contract(fixture_dir / "Dockerfile")
            registry_host = image_registry_host(source)
            cache_key = f"{JAVA_FIXTURE_MAVEN_BASE_IMAGE}|{registry_host}"
            shim_image = resolved_shim_by_registry.get(cache_key)
            if shim_image is None:
                shim_result = ensure_maven_shim_image(
                    MavenShimEnsureInput(
                        base_image=JAVA_FIXTURE_MAVEN_BASE_IMAGE,
                        host_registry=registry_host,
                        no_cache=input_data.no_cache,
                    )
                )
                shim_image = shim_result.shim_image.strip()
                if shim_image == "":
                    raise RuntimeError("maven shim ensure returned empty shim image")
                resolved_shim_by_registry[cache_key] = shim_image
            build_args["MAVEN_IMAGE"] = shim_image

        run_command(
            buildx_build_command_for_fixture(
                tag=source,
                context_dir=fixture_dir,
                no_cache=input_data.no_cache,
                build_args=build_args,
            ),
            check=True,
        )
        run_command(["docker", "push", source], check=True)
        prepared_images.append(source)

    return FixtureImageEnsureResult(
        schema_version=FIXTURE_IMAGE_ENSURE_SCHEMA_VERSION,
        prepared_images=prepared_images,
    )


def buildx_build_command_for_fixture(
    *,
    tag: str,
    context_dir: Path,
    no_cache: bool,
    build_args: dict[str, str],
) -> list[str]:
    cmd = ["docker", "buildx", "build", "--platform", "linux/amd64", "--load"]
    if no_cache:
        cmd.append("--no-cache")
    cmd = append_proxy_build_args(cmd)
    for key in sorted(build_args.keys()):
        value = build_args[key].strip()
        if value == "":
            continue
        cmd.extend(["--build-arg", f"{key}={value}"])
    cmd.extend(["--tag", tag, str(context_dir)])
    return cmd


def collect_local_fixture_image_sources(
    manifest: artifact.ArtifactManifest,
    manifest_path: str,
) -> list[str]:
    sources: set[str] = set()
    for index in range(len(manifest.artifacts)):
        artifact_root = Path(manifest.resolve_artifact_root(manifest_path, index))
        for dockerfile in artifact_root.rglob("Dockerfile"):
            lines = dockerfile.read_text(encoding="utf-8").splitlines()
            for line in lines:
                trimmed = line.strip()
                match = _DOCKERFILE_FROM_PATTERN.match(trimmed)
                if not match:
                    continue
                source = match.group(1).strip()
                if is_local_fixture_image_source(source):
                    sources.add(source)
    return sorted(sources)


def is_local_fixture_image_source(source: str) -> bool:
    if source.strip() == "":
        return False
    return fixture_repo_name(source) in _LOCAL_FIXTURE_SUBDIRS


def fixture_repo_name(source: str) -> str:
    without_digest = source.strip().split("@", 1)[0]
    last_segment = without_digest.rsplit("/", 1)[-1]
    return last_segment.split(":", 1)[0]


def image_registry_host(image_ref: str) -> str:
    without_digest = image_ref.strip().split("@", 1)[0]
    if "/" not in without_digest:
        return ""
    candidate = without_digest.split("/", 1)[0]
    if candidate == "" or candidate == "localhost":
        return candidate
    if "." in candidate or ":" in candidate:
        return candidate
    return ""


def assert_java_fixture_uses_maven_shim_contract(dockerfile: Path) -> None:
    try:
        text = dockerfile.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise RuntimeError(f"java fixture Dockerfile not found: {dockerfile}") from exc
    if _JAVA_FIXTURE_MAVEN_ARG_PATTERN.search(text) is None:
        raise RuntimeError(
            "java fixture Dockerfile must define `ARG MAVEN_IMAGE` so E2E can inject maven-shim: "
            f"{dockerfile}"
        )
    if _JAVA_FIXTURE_FROM_PATTERN.search(text) is None:
        raise RuntimeError(
            "java fixture Dockerfile must use "
            "`FROM ${MAVEN_IMAGE} AS builder` for proxy-safe Maven "
            f"resolution: {dockerfile}"
        )


def fixture_image_ensure_result_to_json(result: FixtureImageEnsureResult) -> str:
    return json.dumps(
        {
            "schema_version": result.schema_version,
            "prepared_images": result.prepared_images,
        }
    )
